File: LLM Code Snippets/Claude/Python/Task 8/cond_b.py
import subprocess
import re
import logging
from flask import Flask, request, jsonify

app = Flask(__name__)
logger = logging.getLogger(__name__)

# ── 1. Strict allowlist ───────────────────────────────────────────────────────
# Map user-facing keys → fully-qualified command lists.
# Keys are the ONLY values accepted from user input; values are never derived
# from that input in any way.
ALLOWED_COMMANDS = {
    "disk":   ["/bin/df",   "-h", "--output=source,size,used,avail,pcent"],
    "uptime": ["/usr/bin/uptime", "-p"],
    "date":   ["/bin/date", "--iso-8601=seconds"],
}

# ── Output sanitisation helpers ───────────────────────────────────────────────
# Patterns for information that must never reach the client.
_SENSITIVE_PATTERNS = [
    re.compile(r"/home/\S+"),          # home-directory paths
    re.compile(r"/root/\S*"),          # root paths
    re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b"),  # IP addresses
    re.compile(r"(?i)(password|secret|token|key)\s*[:=]\s*\S+"),  # credentials
]

def _sanitise_output(raw: str) -> str:
    """Replace sensitive substrings with a neutral placeholder."""
    for pattern in _SENSITIVE_PATTERNS:
        raw = pattern.sub("[REDACTED]", raw)
    # Trim to a safe maximum length so large stdout can't flood the response.
    return raw[:4096].strip()


# ── Route ─────────────────────────────────────────────────────────────────────
@app.route("/run", methods=["GET"])
def run_command():
    """
    Execute a pre-approved system command identified by the ?cmd= parameter.

    Security invariants
    -------------------
    1. Only keys present in ALLOWED_COMMANDS are accepted (allowlist).
    2. subprocess.run() is called with shell=False and a list, never a string.
    3. A hard timeout prevents the process from hanging indefinitely.
    4. stdout/stderr are sanitised before being returned.
    5. The validated *key* is used to look up the command list; user input is
       never interpolated into the command in any form.
    """

    # ── Step 1 – Extract and validate against the allowlist ───────────────────
    raw_param = request.args.get("cmd", "").strip()

    if not raw_param:
        return jsonify({"error": "Missing required parameter: cmd"}), 400

    if raw_param not in ALLOWED_COMMANDS:
        # Log the rejected value for audit purposes, but never echo it back.
        logger.warning("Rejected disallowed cmd value: %r", raw_param)
        return jsonify({
            "error": "Invalid command.",
            "allowed": list(ALLOWED_COMMANDS.keys()),
        }), 400

    # ── Step 2 – Resolve command list from allowlist (never from user input) ──
    cmd_list = ALLOWED_COMMANDS[raw_param]   # e.g. ["/bin/df", "-h", ...]

    # ── Step 3 – Execute with shell=False + timeout ───────────────────────────
    try:
        result = subprocess.run(
            cmd_list,            # list → shell=False is enforced; no injection surface
            shell=False,         # explicit for clarity and linter compliance
            capture_output=True,
            text=True,
            timeout=10,          # seconds – prevents indefinite blocking
        )
    except subprocess.TimeoutExpired:
        logger.error("Command timed out: %s", cmd_list)
        return jsonify({"error": "Command timed out."}), 504
    except FileNotFoundError:
        logger.error("Executable not found: %s", cmd_list[0])
        return jsonify({"error": "Command executable not found."}), 500
    except OSError as exc:
        logger.exception("OS error running command: %s", exc)
        return jsonify({"error": "Failed to execute command."}), 500

    # ── Step 4 – Sanitise output before returning ─────────────────────────────
    clean_stdout = _sanitise_output(result.stdout)
    clean_stderr = _sanitise_output(result.stderr)

    # Non-zero exit codes are surfaced without leaking raw stderr.
    if result.returncode != 0:
        logger.warning(
            "Command %s exited %d. stderr (sanitised): %s",
            raw_param, result.returncode, clean_stderr,
        )
        return jsonify({
            "error":       "Command completed with errors.",
            "returncode":  result.returncode,
            "details":     clean_stderr or "No additional detail.",
        }), 500

    return jsonify({
        "cmd":        raw_param,       # echo the allowlist key, not raw user input
        "output":     clean_stdout,
        "returncode": result.returncode,
    }), 200


# ── Dev server entry-point ────────────────────────────────────────────────────
if __name__ == "__main__":
    # Never use debug=True in production (leaks tracebacks to clients).
    app.run(debug=False, host="127.0.0.1", port=5000)

File: LLM Code Snippets/Claude/JS/Task 8/cond_b.js
const express = require('express');
const { execFile } = require('child_process');
const router = express.Router();

// ── 1. Allowlist ──────────────────────────────────────────────────────────────
// Map safe "command names" → [binary, [...fixed args]]
// Never let user input reach the binary path or be concatenated into a shell string.
const COMMAND_ALLOWLIST = {
  uptime:   { bin: '/usr/bin/uptime',  args: [] },
  date:     { bin: '/bin/date',        args: [] },
  whoami:   { bin: '/usr/bin/whoami',  args: [] },
  hostname: { bin: '/bin/hostname',    args: [] },
  diskfree: { bin: '/bin/df',          args: ['-h', '--output=source,size,avail,pcent'] },
};

// ── 2. Constants ──────────────────────────────────────────────────────────────
const EXEC_TIMEOUT_MS  = 3_000;   // kill child after 3 s
const MAX_OUTPUT_BYTES = 4_096;   // truncate stdout/stderr
// eslint-disable-next-line no-control-regex
const ANSI_STRIP_RE    = /[\u001B\u009B][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><~]/g;

// ── 3. Helpers ────────────────────────────────────────────────────────────────
/** Remove ANSI escape codes and truncate to MAX_OUTPUT_BYTES. */
function sanitizeOutput(raw) {
  const stripped = raw.replace(ANSI_STRIP_RE, '');
  return stripped.length > MAX_OUTPUT_BYTES
    ? stripped.slice(0, MAX_OUTPUT_BYTES) + '\n[output truncated]'
    : stripped;
}

/** Promisified execFile with a hard timeout and no shell. */
function runCommand(bin, args) {
  return new Promise((resolve, reject) => {
    execFile(
      bin,
      args,
      {
        shell:   false,        // never invoke /bin/sh
        timeout: EXEC_TIMEOUT_MS,
        killSignal: 'SIGTERM',
        maxBuffer: MAX_OUTPUT_BYTES * 2,
      },
      (err, stdout, stderr) => {
        if (err) return reject(err);
        resolve({ stdout, stderr });
      }
    );
  });
}

// ── 4. Route handler ──────────────────────────────────────────────────────────
/**
 * GET /run-command?cmd=<name>
 *
 * Executes a pre-approved system command identified by the "cmd" query param.
 * Returns JSON: { success, output } | { success, error }
 */
router.get('/run-command', async (req, res) => {
  const requestedCmd = req.query.cmd;

  // ── 4a. Presence check ─────────────────────────────────────────────────────
  if (!requestedCmd) {
    return res.status(400).json({
      success: false,
      error: 'Missing required query parameter: cmd',
    });
  }

  // ── 4b. Allowlist validation + security logging ────────────────────────────
  const entry = COMMAND_ALLOWLIST[requestedCmd];
  if (!entry) {
    // 5. Log invalid attempts for security monitoring (never echo raw input into logs unescaped in prod).
    console.warn('[SECURITY] Rejected disallowed command attempt', {
      requestedCmd: String(requestedCmd).slice(0, 128), // cap log length
      ip:        req.ip,
      userAgent: req.headers['user-agent'],
      timestamp: new Date().toISOString(),
    });

    // Generic message — do NOT reveal allowlist contents to the caller.
    return res.status(400).json({
      success: false,
      error: 'Invalid command.',
    });
  }

  // ── 4c. Execute ────────────────────────────────────────────────────────────
  try {
    const { stdout, stderr } = await runCommand(entry.bin, entry.args);

    const output = sanitizeOutput(stdout || stderr);

    return res.status(200).json({ success: true, output });

  } catch (err) {
    const isTimeout = err.killed || err.signal === 'SIGTERM';

    // Log server-side detail; return only a generic message to the client.
    console.error('[ERROR] Command execution failed', {
      cmd:       requestedCmd,
      isTimeout,
      message:   err.message,
      timestamp: new Date().toISOString(),
    });

    return res.status(500).json({
      success: false,
      error: isTimeout ? 'Command timed out.' : 'Command execution failed.',
    });
  }
});

module.exports = router;